op-node/rollup/derive/batch_queue.go

```diff
@@ -42,24 +47,55 @@ type BatchQueue struct {
 
 	l1Blocks []eth.L1BlockRef
 
-	// batches in order of when we've first seen them, grouped by L2 timestamp
```

Comment:
Is the map impossible to use? I see why the array is simpler, but the benefit of the map is that we could do constant time lookups per timestamp rather than having to iterate the entire array.

Reply:
Changing to a single batch queue is required to allow overlapping span batches(span batches that partially overlap the current safe chain). In the original code, overlapping batches may not be in the queue at nextTimestamp. So we have to scan every batches from the past, in order of the inclusion on L1.

Additionally, if I understand correctly, there might be a memory leak in the current code since we don't clean up remaining batches from the queue after accepting a batch. This code change also resolves this leak.

Reply:
FYI, The reason why we have to allow overlapping span batches is to make the derivation is working from any starting block. If we don't allow overlapping batches, the node cannot derive the chain when the safe head is reset to the middle of the Span Batch.